File: nexus-h1/integrations/whatsapp.py
#!/usr/bin/env python3
"""
Nexus H1 — WhatsApp Integration Module
Python wrapper around whatsapp-web.js Node.js process
"""

import json
import subprocess
import time
from pathlib import Path
from typing import List, Dict, Optional

JS_PATH = Path(__file__).parent / "whatsapp.js"
NODE_CMD = "node"


class WhatsAppClient:
    def __init__(self):
        self.process: Optional[subprocess.Popen] = None
        self._buffer = ""

    def start(self):
        """Start the WhatsApp client process."""
        self.process = subprocess.Popen(
            [NODE_CMD, str(JS_PATH)],
            stdin=subprocess.PIPE,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True,
            bufsize=1,
        )
        # Wait for ready or QR
        print("Starting WhatsApp client...")
        print("Scan QR code from secrets/whatsapp_qr.txt if prompted")

    def stop(self):
        """Stop the client."""
        if self.process:
            self.process.terminate()
            self.process.wait(timeout=5)
            self.process = None

    def _send_cmd(self, cmd: Dict) -> List[Dict]:
        """Send command and collect response lines."""
        if not self.process:
            raise RuntimeError("WhatsApp client not started")
        
        self.process.stdin.write(json.dumps(cmd) + "\n")
        self.process.stdin.flush()
        
        responses = []
        # Wait up to 10 seconds for response
        for _ in range(100):
            line = self.process.stdout.readline().strip()
            if line:
                responses.append(line)
                if line.startswith("SENT") or line.startswith("CHATS") or line.startswith("MESSAGES") or line.startswith("STATUS") or line.startswith("QUEUE"):
                    break
            time.sleep(0.1)
        return responses

    def send_message(self, to: str, message: str) -> str:
        """Send text message. 'to' should be phone number with country code."""
        responses = self._send_cmd({"action": "send", "to": to, "message": message})
        for r in responses:
            if r.startswith("SENT"):
                return r
        return "ERROR"

    def get_chats(self) -> List[Dict]:
        """Get recent chats."""
        responses = self._send_cmd({"action": "get_chats"})
        for r in responses:
            if r.startswith("CHATS"):
                _, data = r.split(" ", 1)
                return json.loads(data)
        return []

    def get_messages(self, chat_id: str, limit: int = 50) -> List[Dict]:
        """Get messages from a chat."""
        responses = self._send_cmd({"action": "get_messages", "chatId": chat_id, "limit": limit})
        for r in responses:
            if r.startswith("MESSAGES"):
                _, data = r.split(" ", 1)
                return json.loads(data)
        return []

    def get_status(self) -> Dict:
        """Get client status."""
        responses = self._send_cmd({"action": "status"})
        for r in responses:
            if r.startswith("STATUS"):
                _, data = r.split(" ", 1)
                return json.loads(data)
        return {"ready": False}

    def get_pending_messages(self) -> List[Dict]:
        """Get and clear pending incoming messages."""
        responses = self._send_cmd({"action": "get_queue"})
        for r in responses:
            if r.startswith("QUEUE"):
                _, data = r.split(" ", 1)
                return json.loads(data)
        return []


# Convenience functions for direct usage
_client = None

def start():
    global _client
    _client = WhatsAppClient()
    _client.start()

def stop():
    global _client
    if _client:
        _client.stop()
        _client = None

def send(to: str, message: str) -> str:
    if not _client:
        raise RuntimeError("WhatsApp not started. Call start() first.")
    return _client.send_message(to, message)

def chats() -> List[Dict]:
    if not _client:
        raise RuntimeError("WhatsApp not started. Call start() first.")
    return _client.get_chats()

def messages(chat_id: str, limit: int = 50) -> List[Dict]:
    if not _client:
        raise RuntimeError("WhatsApp not started. Call start() first.")
    return _client.get_messages(chat_id, limit)

def pending() -> List[Dict]:
    if not _client:
        raise RuntimeError("WhatsApp not started. Call start() first.")
    return _client.get_pending_messages()

def status() -> Dict:
    if not _client:
        return {"ready": False}
    return _client.get_status()


if __name__ == "__main__":
    import sys
    if len(sys.argv) > 1 and sys.argv[1] == "start":
        start()
        try:
            while True:
                time.sleep(1)
                s = status()
                if s.get("ready"):
                    print("WhatsApp is ready!")
                    break
        except KeyboardInterrupt:
            stop()
    else:
        print("Usage: python whatsapp.py start")

File: nexus-h1/integrations/whatsapp.js
const { Client, LocalAuth } = require('whatsapp-web.js');
const fs = require('fs');
const path = require('path');

const SESSION_PATH = path.join(__dirname, '..', 'secrets', 'whatsapp_session');

const client = new Client({
    authStrategy: new LocalAuth({ dataPath: SESSION_PATH }),
    puppeteer: {
        headless: true,
        args: ['--no-sandbox', '--disable-setuid-sandbox']
    }
});

// Message queue for incoming messages
let messageQueue = [];
let isReady = false;

client.on('qr', (qr) => {
    console.log('QR_RECEIVED');
    console.log(qr);
    // Save QR to file for external scanning
    fs.writeFileSync(path.join(__dirname, '..', 'secrets', 'whatsapp_qr.txt'), qr);
});

client.on('ready', () => {
    isReady = true;
    console.log('WHATSAPP_READY');
});

client.on('message', async (msg) => {
    // Ignore own messages and status broadcasts
    if (msg.fromMe || msg.from === 'status@broadcast') return;
    
    const messageData = {
        id: msg.id.id,
        from: msg.from,
        fromMe: msg.fromMe,
        body: msg.body,
        timestamp: msg.timestamp,
        type: msg.type,
        hasMedia: msg.hasMedia,
        author: msg.author || msg.from
    };
    
    messageQueue.push(messageData);
    
    // Also append to log file
    const logEntry = `[${new Date().toISOString()}] ${msg.from}: ${msg.body}\n`;
    fs.appendFileSync(path.join(__dirname, '..', 'memory', 'whatsapp_messages.log'), logEntry);
    
    console.log('MESSAGE_RECEIVED', JSON.stringify(messageData));
});

client.on('disconnected', (reason) => {
    isReady = false;
    console.log('WHATSAPP_DISCONNECTED', reason);
});

// Command interface via stdin
process.stdin.on('data', async (data) => {
    const line = data.toString().trim();
    if (!line) return;
    
    try {
        const cmd = JSON.parse(line);
        
        switch (cmd.action) {
            case 'send': {
                const { to, message } = cmd;
                const chatId = to.includes('@c.us') ? to : `${to}@c.us`;
                const sent = await client.sendMessage(chatId, message);
                console.log('SENT', JSON.stringify({ id: sent.id.id, to }));
                break;
            }
            case 'send_media': {
                const { to, path: mediaPath, caption } = cmd;
                const chatId = to.includes('@c.us') ? to : `${to}@c.us`;
                const media = require('whatsapp-web.js').MessageMedia.fromFilePath(mediaPath);
                const sent = await client.sendMessage(chatId, media, { caption });
                console.log('SENT_MEDIA', JSON.stringify({ id: sent.id.id, to }));
                break;
            }
            case 'get_chats': {
                const chats = await client.getChats();
                const simplified = chats.slice(0, 50).map(c => ({
                    id: c.id._serialized,
                    name: c.name,
                    isGroup: c.isGroup,
                    unreadCount: c.unreadCount,
                    timestamp: c.timestamp
                }));
                console.log('CHATS', JSON.stringify(simplified));
                break;
            }
            case 'get_messages': {
                const { chatId, limit = 50 } = cmd;
                const chat = await client.getChatById(chatId);
                const messages = await chat.fetchMessages({ limit });
                const simplified = messages.map(m => ({
                    id: m.id.id,
                    from: m.from,
                    body: m.body,
                    timestamp: m.timestamp,
                    fromMe: m.fromMe
                }));
                console.log('MESSAGES', JSON.stringify(simplified));
                break;
            }
            case 'status': {
                console.log('STATUS', JSON.stringify({ ready: isReady, queueSize: messageQueue.length }));
                break;
            }
            case 'get_queue': {
                console.log('QUEUE', JSON.stringify(messageQueue));
                messageQueue = []; // clear after reading
                break;
            }
            default:
                console.log('ERROR', 'Unknown action:', cmd.action);
        }
    } catch (err) {
        console.log('ERROR', err.message);
    }
});

client.initialize();
